mcf-browser/src/back-end/graph.js

```diff
@@ -0,0 +1,318 @@
+    this.dcid = null;
+    this.assertions = null;
+    this.invAssertions = null;
+  }
+
+  /**
+   * Indicates if a given object is an instance of Node class.
+   * @param {Object} obj The object to check.
+   * @return {boolean} True if the object is an instance of Node.
+   */
+  static isNode(obj) { return obj instanceof Node; }
+
+  /**
+   * Returns a node with the given ID, creates a new node if shouldCreate is
+   * true, or null if the node does not exist and a node should not be created.
+   *
+   * @param {string} id The id of the node to find.
+   * @param {boolean} shouldCreate True if a new Node should be created if it
+   *     does not already exist.
+   * @return {Node|null} The found node if it exists or is created.
+   */
+  static getNode(id, shouldCreate) {
+    const existing = Node.nodeHash[id];
+    if (existing) {
+      return existing;
+    }
+
+    if (shouldCreate) {
+      const newNode = new Node(id);
+      Node.nodeHash[id] = newNode;
+      return newNode;
+    }
+    return null;
+  }
+
+  /**
+   * Returns the reference to the node that is displayed in browser. If the
+   * node has a dcid, then the dcid will be displayed. If the node's local id
+   * is different, then the local id is also displayed.
+   * Ex: <dcid> [l:<localId>]
+   * @return {string} The reference to the node to be displayed.
+   */
+  getRef() {
+    const dcidRef = this.dcid ? this.dcid : '';
+    let localRef = '';
+
+    if (!this.dcid || this.dcid !== this.localId) {
+      localRef = '[l:' + this.localId + ']';
+    }
+    return [ dcidRef, localRef ].join(' ').trim();
+  }
+
+  /**
+   * Sets the dcid of Node object. Checks if a separate node based on the dcid
+   * already exists. If remote node exists, then the remote node is absorbed by
+   * current node via mergeNode() method.
+   *
+   * @param {string} dcid The dcid to be added to the Node object.
+   */
+  setDCID(dcid) {
+    const remote = Node.nodeHash[dcid];
+    if (remote) {
+      this.mergeNode(remote);
+    }
+    this.dcid = dcid;
+    Node.nodeHash[dcid] = this;
+  }
+
+  /**
+   * Moves the assertions and inverse Assertions from the given param node to
+   * the calling Node object by changing the src property for assertions and the
+   * target property of the invAssertions.
+   *
+   * @param {Node} absorbedNode The node object whose triples should be copied.
+   */
+  mergeNode(absorbedNode) {
+    if (this.localId === absorbedNode.localId) {
+      return;
+    }
+
+    absorbedNode.getAssertions().forEach((assert) => {
+      assert.src = this;
+      assert.nextAssertion = this.assertions;
+      this.assertions = assert;
+    });
+
+    absorbedNode.getInvAssertions().forEach((invAssert) => {
+      invAssert.target = this;
+      invAssert.invNextAssertion = this.invAssertions;
+      this.invAssertions = invAssert;
+    });
+  }
+
+  /**
+   * Sets the property existsInKG to true if the Node has triples in the DC KG.
+   */
+  async setExistsInKG() {
```

Comment:
Who calls this?

Reply:
This is called from the 'front-end" side. I have set up an in between file called server-api.js which is the only file that the 'front-end' files use for access to back-end functions. This server-api.js file serves as wrapper for all of the back-end functionalities, including calling this function. I have not submitted a PR for that code yet.